core/flyout_base.js

```diff
@@ -426,77 +426,60 @@ Blockly.Flyout.prototype.hide = function() {
-        xmlList);
+        contentInfo);
     if (typeof fnToApply != 'function') {
       throw TypeError('Couldn\'t find a callback function when opening' +
           ' a toolbox category.');
     }
-    xmlList = fnToApply(this.workspace_.targetWorkspace);
-    if (!Array.isArray(xmlList)) {
+    contentInfo = fnToApply(this.workspace_.targetWorkspace);
+    if (!Array.isArray(contentInfo)) {
       throw TypeError('Result of toolbox category callback must be an array.');
     }
   }
+  // contentInfo is either an array with JSON, an Array with xml, or a NodeList.
+  var parsedContent = Blockly.utils.toolbox.parseToolbox(contentInfo);
 
   this.setVisible(true);
   // Create the blocks to be shown in this flyout.
   var contents = [];
   var gaps = [];
   this.permanentlyDisabled_.length = 0;
-  var default_gap = this.horizontalLayout_ ? this.GAP_X : this.GAP_Y;
-  for (var i = 0, xml; (xml = xmlList[i]); i++) {
-    if (!xml.tagName) {
+  var defaultGap = this.horizontalLayout_ ? this.GAP_X : this.GAP_Y;
+  for (var i = 0, contentItem; (contentItem = parsedContent[i]); i++) {
+    if (!contentItem.contenttype) {
       continue;
     }
-    switch (xml.tagName.toUpperCase()) {
+    switch (contentItem.contenttype.toUpperCase()) {
       case 'BLOCK':
-        var curBlock = Blockly.Xml.domToBlock(xml, this.workspace_);
-        if (!curBlock.isEnabled()) {
-          // Record blocks that were initially disabled.
-          // Do not enable these blocks as a result of capacity filtering.
-          this.permanentlyDisabled_.push(curBlock);
-        }
-        contents.push({type: 'block', block: curBlock});
-        // This is a deprecated method for adding gap to a block.
-        // <block type="math_arithmetic" gap="8"></block>
-        var gap = parseInt(xml.getAttribute('gap'), 10);
-        gaps.push(isNaN(gap) ? default_gap : gap);
+        var blockInfo = /** @type {Blockly.utils.toolbox.BlockInfo} */ (contentItem);
+        this.addBlock_(blockInfo, contents, gaps, defaultGap);
         break;
       case 'SEP':
-        // Change the gap between two toolbox elements.
-        // <sep gap="36"></sep>
-        // The default gap is 24, can be set larger or smaller.
-        // This overwrites the gap attribute on the previous element.
-        var newGap = parseInt(xml.getAttribute('gap'), 10);
-        // Ignore gaps before the first block.
-        if (!isNaN(newGap) && gaps.length > 0) {
-          gaps[gaps.length - 1] = newGap;
-        } else {
-          gaps.push(default_gap);
-        }
+        var sepInfo = /** @type {Blockly.utils.toolbox.SeparatorInfo} */ (contentItem);
+        this.addSeparator_(sepInfo, gaps, defaultGap);
         break;
       case 'LABEL':
+        var labelInfo = /** @type {Blockly.utils.toolbox.LabelInfo} */ (contentItem);
+        // A label is a button with different styling.
+        this.addButton_(labelInfo, contents, gaps, defaultGap, true);
```

Comment:
Any reason we couldn't pass an object to that consists of: {labelInfo, contents, gaps, etc.. } rather than passing all these parameters.